### PR TITLE
Refactor WFSS transforms

### PR DESCRIPTION
**Pull request by nden**
Improved performance.

This PR uses the `expected values` in the wfss tests (assign_wcs/tests/test_wfss.py) but changes the calls to the functions to use the new interface.

### CODE REVIEW

File: jwst/assign_wcs/tests/test_wfss.py [hunk 1]
```diff
@@ -235,21 +224,21 @@ def test_NIRISSBackwardDispersion():
          [15.66666438, 15.6661858 , 15.66570726, 15.66522876, 15.66475031, 15.66427189, 15.66379352]])
 
     # refactored call
-    # x, y = wcstools.grid_from_bounding_box(slit.meta.wcs.bounding_box)
-    # xdx, ydy, _, _ = model(x, y, wavelength, np.zeros(x.shape)+1)
+    x, y = grid_from_bounding_box(slit.meta.wcs.bounding_box)
+    xdx, ydy, _, _, _ = model(x, y, wavelength, np.zeros(x.shape)+1)
```
Comment (hbushouse) [inline jwst/assign_wcs/tests/test_wfss.py]: Up above in line 150 you changed this call to use `order` in place of the hardcoded value "1". Is the same change warranted here?

File: jwst/extract_2d/grisms.py [hunk 2]
```diff
@@ -551,11 +551,13 @@ def compute_wavelength_array(slit):
...
+    #shape = grid[0].shape
+    #wavelength = np.empty(shape, dtype=np.float64)
     transform = slit.meta.wcs.forward_transform
-    for j in range(shape[0]):
-        for i in range(shape[1]):
-            wavelength[j, i] = transform(grid[0][j, i], grid[1][j, i])[2]
+    #for j in range(shape[0]):
+        #for i in range(shape[1]):
+            #wavelength[j, i] = transform(grid[0][j, i], grid[1][j, i])[2]
+    x, y = grid_from_bounding_box(slit.meta.wcs.bounding_box)
```
Comment (hbushouse) [inline jwst/extract_2d/grisms.py]: Woot!!! No more looping over pixels!

File: jwst/transforms/models.py [hunk 3]
```diff
@@ -1406,25 +1444,44 @@ def evaluate(self, x, y, wavelength, order):
...
-        dy = self.ymodels[iorder][0](x, y) + t * self.ymodels[iorder][1](x, y)
-        # rotate by theta
+        xmodel = self.xmodels[iorder]
+        ymodel = self.ymodels[iorder]
+        lmodel = self.lmodels[iorder]
+
+        dx = xmodel[0](x, y) + t * xmodel[1](x, y)
+        dy = ymodel[0](x, y) + t * ymodel[1](x, y)
+
+        ## use that t to compute the dx and dy
```
Comment (hbushouse) [inline jwst/transforms/models.py]: Are all of these commented blocks new code that will be activated in the future?

**Comment by nden** [comment]
One of the tests is failing the comparison with the previous version of the transforms. The refactored code uses an interpolation routine from scipy while the previous code used the one in numpy setting  wavelength values outside the bounding_box to the left and right corner. The current one cannot do this. Values outside the bounding box can be set to either NaN or extrapolated. I'd argue one of these is better than setting the wavelength to a known incorrect value.

File: jwst/assign_wcs/tests/test_wfss.py (hunk 1 above)
Comment (nden) [reply jwst/assign_wcs/tests/test_wfss.py]: Yes, I'll change it. I'm running the tests for order=1 only.

File: jwst/transforms/models.py (hunk 3 above)
Comment (nden) [reply jwst/transforms/models.py]: Everything is activated now, I will clean the code.

**Comment by hbushouse** [comment]
> One of the tests is failing the comparison with the previous version of the transforms. The refactored code uses an interpolation routine from scipy while the previous code used the one in numpy setting wavelength values outside the bounding_box to the left and right corner. The current one cannot do this. Values outside the bounding box can be set to either NaN or extrapolated. I'd argue one of these is better than setting the wavelength to a known incorrect value.

Seems to me we have (or had) places downstream where wavelength values are set to NaN if they're outside of the bounding_box or are otherwise unable to be calculated for any reason, so there's precedence for this approach. @hodge can you confirm? I would vote for setting them to NaN, rather than setting them a bad value.

**Comment by philhodge** [comment]
There are some places where we assume that wavelengths (or right ascension or declination) will be NaN for pixels that are outside the bounding box.  It's useful to have an unambiguous and simple test for points that are outside the bounding box.  In `photom`, we temporarily set elements of the conversion factor to NaN for pixels that have NaN wavelengths, then check for NaN in the array of conversion factors and replace those with zero.

**Pull request merged**